### PR TITLE
micropython support for Sound

### PR DESCRIPTION
**Pull request by dwalton76**
For #628 

Tested micropython and python3 via
```
from ev3dev2.sound import Sound

VOLUME = 30

sound = Sound()
sound.speak("hello world", volume=VOLUME, play_type=Sound.PLAY_WAIT_FOR_COMPLETE)

sound.play_song((
    ('Db4', 'e3'),
))

# Star Wars via play_song()
sound.play_song((
    ('D4', 'e3'),      # intro anacrouse
    ('D4', 'e3'),
    ('D4', 'e3'),
    ('G4', 'h'),       # meas 1
    ('D5', 'h'),
    ('C5', 'e3'),      # meas 2
    ('B4', 'e3'),
    ('A4', 'e3'),
    ('G5', 'h'),
    ('D5', 'q'),
    ('C5', 'e3'),      # meas 3
    ('B4', 'e3'),
    ('A4', 'e3'),
    ('G5', 'h'),
    ('D5', 'q'),
    ('C5', 'e3'),      # meas 4
    ('B4', 'e3'),
    ('C5', 'e3'),
    ('A4', 'h.'),
))

# Emperial March via tone()
sound.tone([
    (392, 350, 100), (392, 350, 100), (392, 350, 100), (311.1, 250, 100),
    (466.2, 25, 100), (392, 350, 100), (311.1, 250, 100), (466.2, 25, 100),
    (392, 700, 100), (587.32, 350, 100), (587.32, 350, 100),
    (587.32, 350, 100), (622.26, 250, 100), (466.2, 25, 100),
    (369.99, 350, 100), (311.1, 250, 100), (466.2, 25, 100), (392, 700, 100),
    (784, 350, 100), (392, 250, 100), (392, 25, 100), (784, 350, 100),
    (739.98, 250, 100), (698.46, 25, 100), (659.26, 25, 100),
    (622.26, 25, 100), (659.26, 50, 400), (415.3, 25, 200), (554.36, 350, 100),
    (523.25, 250, 100), (493.88, 25, 100), (466.16, 25, 100), (440, 25, 100),
    (466.16, 50, 400), (311.13, 25, 200), (369.99, 350, 100),
    (311.13, 250, 100), (392, 25, 100), (466.16, 350, 100), (392, 250, 100),
    (466.16, 25, 100), (587.32, 700, 100), (784, 350, 100), (392, 250, 100),
    (392, 25, 100), (784, 350, 100), (739.98, 250, 100), (698.46, 25, 100),
    (659.26, 25, 100), (622.26, 25, 100), (659.26, 50, 400), (415.3, 25, 200),
    (554.36, 350, 100), (523.25, 250, 100), (493.88, 25, 100),
    (466.16, 25, 100), (440, 25, 100), (466.16, 50, 400), (311.13, 25, 200),
    (392, 350, 100), (311.13, 250, 100), (466.16, 25, 100),
    (392.00, 300, 150), (311.13, 250, 100), (466.16, 25, 100), (392, 700)
])

sound.play_file("/usr/share/sounds/ev3dev/mechanical/airbrake.wav", volume=VOLUME, play_type=Sound.PLAY_NO_WAIT_FOR_COMPLETE)
```

### CODE REVIEW

**Comment by WasabiFan** [comment]
This changes the API, right? It removes one of the "play type" options and removes the handle returns?

**Comment by dwalton76** [comment]
All of the `play_type` options are still the same.  The return value has changed if you are doing `play_type=Sound.PLAY_NO_WAIT_FOR_COMPLETE`...that used to return the PID of the subprocess but now we just return `None`.  Our default is `Sound.PLAY_WAIT_FOR_COMPLETE`...my gut says there probably aren't a ton of folks using the `NO_WAIT` option and doing much with that returned PID.  With the proposed change they could always figure out the PID by looking at the processes that are currently running.

**Comment by WasabiFan** [comment]
Ok, so am I correct in thinking that with this change there isn't an API to play a sound asynchronously (i.e., start it then keep going while it plays)? I feel like that behavior could be useful, as it might be nice to speak/beep some status while your robot is still going.

**Comment by dwalton76** [comment]
No, that functionality is still there, to do that you use `play_type=Sound.PLAY_NO_WAIT_FOR_COMPLETE`.

**Comment by WasabiFan** [comment]
I'm still a bit concerned about the breaking change on the return type. Perhaps this is unfounded, but it feels like this might cause problems for existing consumers who do something more advanced.

I think we could instead split the `Popen` stuff into its own method (i.e., just have it as a generic "run audio command" method) and then have it fall back to `os.system` when needed. That would probably be less of a code change, would preserve the existing behavior, and introduce support for the additional platform.

I think this would probably work:

```python
def _audio_command(self, command, play_type=PLAY_WAIT_FOR_COMPLETE):
    if is_micropython():
        if play_type == Sound.PLAY_WAIT_FOR_COMPLETE:
            os.system(command)
        else:
            os.system('{} &'.format(command))
        
        return None
    else:
        with open(os.devnull, 'w') as n:
            subprocess = Popen(shlex.split(command), stdout=n)
            if play_type == Sound.PLAY_WAIT_FOR_COMPLETE:
                subprocess.wait()
                return None
            else:
                return subprocess
```

It should be fairly straightforward to rewrite the other functions in terms of this, and that would get all the behavior we need.

**Comment by dwalton76** [comment]
I think this is backwards compatible now.  The only quirk is that the `return` behavior documented in the docstrings is incorrect for micropython...it will not return the PID of the subprocess.  Do we want to document python3 vs micropython behavior?  I don't know that we do that elsewhere.

File: ev3dev2/sound.py [hunk 1]
```diff
@@ -44,15 +47,29 @@ def _make_scales(notes):
     return res
 
 
+
+def get_command_processes(command):
```
Comment (WasabiFan) [inline ev3dev2/sound.py:51]: Ah, got it; this is for the speak method which uses pipes?

File: ev3dev2/sound.py [hunk 2]
```diff
@@ -370,10 +388,10 @@ def get_volume(self, channel=None):
         if channel is None:
             channel = self._get_channel()
 
-        out = check_output(['amixer', 'get', channel]).decode()
-        m = re.search(r'\[(?P<volume>\d+)%\]', out)
+        out = os.popen(['/usr/bin/amixer', 'get', channel]).read()
```
Comment (WasabiFan) [inline ev3dev2/sound.py:396]: this is supported on Micropython as well?

**Review by WasabiFan — APPROVED** [review]
Nice! This seems reasonable to me; I think it would make the most sense to add a note to those doc comments saying that it doesn't return the object on Micropython.

File: ev3dev2/sound.py (hunk 1 above)
Comment (dwalton76) [reply ev3dev2/sound.py:51]: yep

File: ev3dev2/sound.py (hunk 2 above)
Comment (dwalton76) [reply ev3dev2/sound.py:396]: yep, I tested every Sound feature I could find and they all work with python3 and micropython

**Comment by dwalton76** [comment]
Will update the docstrings to mention about the return values when using python3 vs micropython and then merge this one.

**Pull request merged**